replace Page structure with structure (#1443)
iTwin/stratakit: apps/website/src/content/docs/guides/structure.md

---
title: Structure
description: How to successfully organize your content and functionality
---

> Information without structure is just data. High quality interfaces require logical organization. But structure is perceived and traversed in different ways. The interface one user sees on screen is also parsed by screen reader software, and analyzed by AI agents.

Reviewer: not just screen readers.

```suggestion
Information without structure is just data. High quality interfaces require logical organization. But structure is perceived and traversed in different ways. The interface one user sees on screen is also parsed by assistive technologies, and analyzed by AI agents.
```

It’s important everyone—and everything—experiences the same interface. Fortunately, standard HTML enshrines certain structural rules and cues that benefit all parties. This guide will help you communicate structural relationships in an equitable way.

## Order

An early design for an interface typically exists in a flat, graphical form. Any structural relationships we perceive are spatial only.

The perceived order of two objects is set by visually placing one after the other. This visual order may contradict the order of the objects in the accompanying layers panel. Since the output from the design software is a flattened graphic, this hardly matters.

> HTML is different. The code doesn’t _output_ the interface; it _is_ the interface. This means the structure of the code is paramount. If the visual placement of objects contradicts their order in the source, a sighted user experiences something different from a blind screen reader user, indexing crawler, or AI agent.

Reviewer: I don't like the juxtaposition of blind people with these machines. The intro already covers machines, so we can avoid it here.

```suggestion
HTML is different. The code doesn’t _output_ the interface; it _is_ the interface. This means the structure of the code is paramount. If the visual placement of objects contradicts their order in the source, a sighted user experiences something different from a blind screen reader user.
```

Author: @mayank99 I don't think this would lead anyone to conflate them, but I like the tersity.

In the visual interface, the order of any two objects can appear in three dimensions:

1. Left and right 
2. Above and below
3. Back and front (depth)

Whereas any two components in the source can only be _first and last_:

```jsx
<div>
  <Typography>First object</Typography>
  <Typography>Another object</Typography>
</div>
<Typography>Last object</Typography>
```

Avoid using **CSS** such that it creates a mismatch between visual placement and source order. Use this table to check:

| Where objects are  | Components must be |
| ------------------ | ------------------ |
| left and right     | first and last     |
| above and below    | first and last     |
| back and front     | first and last     |

### Focus order

Interactive elements must be focusable so that the interface can be operated by keyboard. The order in which these elements are encountered and focused must be logical:

| Where buttons are  | Focus must occur  |
| ------------------ | ----------------- |
| left and right     | first then last   |
| above and below    | first then last   |
| back and front     | first then last   |

> Avoid _prioritizing_ the focus of certain elements by applying a positive `tabindex` value or the `autofocus` attribute.

Reviewer: From [earlier comment](https://github.com/iTwin/stratakit/pull/1443/changes#r3202158130):

> There can be legitimate use cases for setting initial focus. Our guidance should take that into consideration.

An example: when a dialog or popover is opened, one might want to automatically focus a textfield in it, or the close button.

See iTwinUI Popover docs for comparison: https://itwinui.bentley.com/docs/popover#accessibility

Author: @mayank99 What I mean is that the consumer (the reader) should not implement `autofocus`. Generally, we are managing focus for them, where it should be managed. But I guess they can use `autofocus` in a dialog and override the default behavior? In any case, I've separated it into a cautionary note.

https://github.com/iTwin/stratakit/pull/1443/commits/567841692c0a3e33c2d0868139c5f13195623a90

Reviewer: Yes, I'm referring to the cases where consumers may need to manually set initial focus.

We can add more specific guidance for this in the future.

Author: @mayank99 It's not clear what you think needs to be addressed here, if anything. I don't have time to guess, so maybe you can make an edit if you are able? I believe https://github.com/iTwin/stratakit/commit/567841692c0a3e33c2d0868139c5f13195623a90 covers everything sufficiently.

Reviewer: It is not sufficient, but works for now. As my last comment mentions, we can punt more detailed additions for future. This thread can remain open after merge, for better visibility.

~~This PR is blocked only by https://github.com/iTwin/stratakit/pull/1443#discussion_r3335498662 (update branch + format)~~. Edit: Fixed in [`15786d2`](https://github.com/iTwin/stratakit/pull/1443/commits/15786d223f2710d9440a637da424645e8ba37b35).

```jsx
/* ❌ */
<Link href="/some/page">Focused last</Link>
<TextField label="Focused first" autoFocus />

> <!-- ❌ -->

Reviewer: Updating the comment for consistency.

```suggestion
/* ❌ */
```

<ButtonGroup aria-label="Button group">
  <Button>Focused second</Button>

>   <Button tabIndex="1">Focused first</Button>

Reviewer: In JavaScript, it's a number, not a string.

```suggestion
  <Button tabIndex={1}>Focused first</Button>
```

As with order, a relationship of belonging needs to be programmatically determinable in the code, not just illustrated visually. 

Nesting HTML elements (placing them inside one another) is an everyday practice. But not all HTML elements contribute to communicating a nested structure. Consider the following example:

```jsx
<div class="ButtonGroup">
  <Button>Bold</Button>
  <Button>Italic</Button>
  <Button>Underline</Button>
</div>
```

The intent to communicate belonging is clear from the nested structure and the `ButtonGroup` classname. It’s also expressed visually, via CSS, by making the buttons share a common border. However, neither a screen reader nor a crawler will acknowledge the `<div>`—and for good reason! We use `<div>`s to organize our HTML, not the information within it.

To communicate the presence of a group, we can encode it with [`role="group"`](https://developer.mozilla.org/en-US/Web/Accessibility/ARIA/Reference/Roles/group_role). Whatever you call this group should be the wording of its accompanying label.

```jsx
<div role="group" aria-label="Text formatting">
  <Button>Bold</Button>
  <Button>Italic</Button>
  <Button>Underline</Button>
</div>
```

> In the [**ButtonGroup**](/components/buttongroup) component, **StrataKit** applies `role="group"` automatically. Since the label will be product-specific, you must apply this as the implementer.

Reviewer: "implementer" is vague (we implement the system, they use it). 

"consumer" is the term we more commonly use.

```suggestion
In the [**ButtonGroup**](/components/buttongroup) component, **StrataKit** applies `role="group"` automatically. Since the label will be product-specific, you must apply this as the consumer.
```

It’s important everyone—and everything—experiences the same interface. Fortunately, standard HTML enshrines certain structural rules and cues that benefit all parties. This guide will help you communicate structural relationships in an equitable way.

## Order

An early design for an interface typically exists in a flat, graphical form. Any structural relationships we perceive are spatial only.

The perceived order of two objects is set by visually placing one after the other. This visual order may contradict the order of the objects in the accompanying layers panel. Since the output from the design software is a flattened graphic, this hardly matters.

HTML is different. The code doesn’t _output_ the interface; it _is_ the interface. This means the structure of the code is paramount. If the visual placement of objects contradicts their order in the source, a sighted user experiences something different from a blind screen reader user.

In the visual interface, the order of any two objects can appear in three dimensions:

> 1. Left and right 

Reviewer: Please keep the file formatted, by installing the recommended extensions (prettier, especially), or by using the pre-commit hook, or by manually running `pnpm format` from the project root.

Author: @mayank99 I did this and it didn't apply any changes, so I assume  (with all the other edits complete) we're good to go!

Reviewer: Nope, still [failing](https://github.com/iTwin/stratakit/actions/runs/26765527572/job/78890773769?pr=1443#step:9:12).

Your branch is out-of-date, so you'll want to merge with `main` and then run `pnpm format` again.

Reviewer: Formatted in [`15786d2`](https://github.com/iTwin/stratakit/pull/1443/commits/15786d223f2710d9440a637da424645e8ba37b35).

2. Above and below
3. Back and front (depth)

Whereas any two components in the source can only be _first and last_:

```jsx
<div>
  <Typography>First object</Typography>
  <Typography>Another object</Typography>
</div>

> <Typography>Last object</Typography>

Reviewer 2: Let's avoid component usage when talking about general HTML structure
```suggestion
<div>
  <span>First object</span>
  <span>Another object</span>
</div>
<span>Last object</span>
```

```

Avoid using **CSS** such that it creates a mismatch between visual placement and source order. Use this table to check:

| Where objects are  | Components must be |
| ------------------ | ------------------ |
| left and right     | first and last     |
| above and below    | first and last     |
| back and front     | first and last     |

### Focus order

Interactive elements must be focusable so that the interface can be operated by keyboard. The order in which these elements are encountered and focused must be logical:

| Where buttons are  | Focus must occur  |
| ------------------ | ----------------- |
| left and right     | first then last   |
| above and below    | first then last   |
| back and front     | first then last   |

Avoid prioritizing the focus of certain elements by applying a positive `tabindex` value. This disrupts an otherwise logical order of focus.

```jsx
/* ❌ */
<ButtonGroup aria-label="Button group">
  <Button>Focused second</Button>
  <Button tabIndex={1}>Focused first</Button>
  <Button>Focused last</Button>

> </ButtonGroup>

Reviewer 2: We introduce the `role="group"` in the "Belonging" section. Maybe the wrapper of this snippet could be a simple div w/o attributes?

```suggestion
/* ❌ */
<div role="group" aria-label="Button group">
  <button>Focused second</button>
  <button tabIndex={1}>Focused first</button>
  <button>Focused last</button>
</div>
```

```

:::caution[The `autofocus` attribute]

The `autofocus` attribute focuses an element when an interface first renders. Automatically focusing a button inside a newly opened dialog or popover may be helpful. However, applying `autofocus` to an element on _page load_ should generally be avoided.

:::

> More complex interactive components like [**Menu**](/components/menu) and [**Dialog**](/components/dialog) manage focus automatically. Implement these components to avoid focus order issues that could trip up a user or agent.

Reviewer: these components are already implemented

```suggestion
More complex interactive components like [**Menu**](/components/menu) and [**Dialog**](/components/dialog) manage focus automatically. Use these components to avoid focus order issues that could trip up a user or agent.
```

Typically, skip links are hidden until focused by keyboard. They are not needed by or available to mouse users.

If you are implementing your own skip link, point the skip link's `href` to the `<main>` element page fragment. To ensure keyboard focus follows the link, include `tabindex="-1"` on the target element.

```html
<a href="#main">skip to content</a>
<!-- page preamble here -->
<main id="main" tabindex="-1">...</main>
```

## Belonging

The Gestalt principle of [common region](https://www.nngroup.com/articles/common-region/) states that _“items within a boundary are perceived as a group”_. From forms to dialogs, all sorts of elements take a border, background color, or both, to communicate their contents _belong together_. 

As with order, a relationship of belonging needs to be programmatically determinable in the code, not just illustrated visually. 

Nesting HTML elements (placing them inside one another) is an everyday practice. But not all HTML elements contribute to communicating a nested structure. Consider the following example:

```jsx

> <div class="ButtonGroup">
>   <Button>Bold</Button>
>   <Button>Italic</Button>
>   <Button>Underline</Button>
> </div>

Reviewer 2: ```suggestion
<div class="ButtonGroup">
  <button>Bold</button>
  <button>Italic</button>
  <button>Underline</button>
</div>
```

```

The intent to communicate belonging is clear from the nested structure and the `ButtonGroup` classname. It’s also expressed visually, via CSS, by making the buttons share a common border. However, neither a screen reader nor a crawler will acknowledge the `<div>`—and for good reason! We use `<div>`s to organize our HTML, not the information within it.

To communicate the presence of a group, we can encode it with [`role="group"`](https://developer.mozilla.org/en-US/Web/Accessibility/ARIA/Reference/Roles/group_role). Whatever you call this group should be the wording of its accompanying label.

```jsx

> <div role="group" aria-label="Text formatting">
>   <Button>Bold</Button>
>   <Button>Italic</Button>
>   <Button>Underline</Button>
> </div>

Reviewer 2: ```suggestion
<div role="group" aria-label="Text formatting">
  <button>Bold</button>
  <button>Italic</button>
  <button>Underline</button>
</div>
```

<Typography>Last object</Typography>
```

Avoid using **CSS** such that it creates a mismatch between visual placement and source order. Use this table to check:

| Where objects are  | Components must be |
| ------------------ | ------------------ |
| left and right     | first and last     |
| above and below    | first and last     |
| back and front     | first and last     |

### Focus order

Interactive elements must be focusable so that the interface can be operated by keyboard. The order in which these elements are encountered and focused must be logical:

| Where buttons are  | Focus must occur  |
| ------------------ | ----------------- |
| left and right     | first then last   |
| above and below    | first then last   |
| back and front     | first then last   |

Avoid _prioritizing_ the focus of certain elements by applying a positive `tabindex` value or the `autofocus` attribute.

```jsx
/* ❌ */
<Link href="/some/page">Focused last</Link>
<TextField label="Focused first" autoFocus />

<!-- ❌ -->
<ButtonGroup aria-label="Button group">
  <Button>Focused second</Button>
  <Button tabIndex="1">Focused first</Button>
  <Button>Focused last</Button>
</ButtonGroup>
```

> More complex interactive components like [**Menu**](/docs/components/menu) and [**Dialog**](/docs/components/dialog) manage focus automatically. Implement these components to avoid focus order issues that could trip up a user or agent.

Reviewer: `/docs` prefix is automatically added by the build system.

```suggestion
More complex interactive components like [**Menu**](/components/menu) and [**Dialog**](/components/dialog) manage focus automatically. Implement these components to avoid focus order issues that could trip up a user or agent.
```

Please find-and-remove all instances of `/docs` in this page.

Author: https://github.com/iTwin/stratakit/pull/1443/commits/0da88bf7eb79754c719363beb9ab48824fe175be

Reviewer: In [`0da88bf`](https://github.com/iTwin/stratakit/pull/1443/commits/0da88bf7eb79754c719363beb9ab48824fe175be), the MDN links were also also accidentally changed.

Author: https://github.com/iTwin/stratakit/pull/1443/commits/186e3eb93a2aa22efbbb0ad202ba6cfd50d07047

    <td></td>
  </tr>
  <tr>
    <td>Figure</td>
    <td>Providing captions for images and other media.</td>
    <td><code>&lt;figure></code> and <code>&lt;figcaption></code>. The <code>&lt;figcaption></code> must appear alongside the media, each inside the <code>&lt;figure></code></td>
    <td>None currently provided</td>
  </tr>
</table>

### Landmarks

[Landmarks](https://www.w3.org/WAI/ARIA/apg/practices/landmark-regions/) are the continents in the map of your interface. Landmarks are large and few in number. Everything (all components) must belong to one landmark or another. They have specific purposes:

- [`<main>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/main): Include just once per screen/page. This encapsulates the main, unique content of the screen.
- `<header>`: Include just one at the start of the document. This may contain a principle `<nav>` element.
- `<nav>`: Multiple can be used for different purposes, such as site navigation, pagination, and tables of content. Give each `<nav>` a unique label, like `<nav aria-label="table of contents">`. This differentiates the landmarks when aggregated in screen reader software.

> - `role="search"`: Identifies the element containing the page’s principle search functionality. Apply the `role` attribution to an element containing the `<form>` element. Do not place `role="search"` on the `<form>` element itself.

Reviewer: We can use the `<search>` element now. (It was still too early when the Page structure guide was originally written)

```suggestion
- [`<search>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/search): Identifies the element containing the page’s principle search functionality.
```

Author: @mayank99 Sweet!

```jsx
<h2>Introduction of a subsection</h2>
<p>The content belonging to the subsection.</p>
```

The provided heading elements (`<h1>` to `<h6>`) are numbered according to section depth. 

- `<h1>` (`variant="h1"`): introduces the page
  - `<h2>` (`variant="h2"`): introduces a subsection in the page
  - `<h2>`: there may be multiple subsections of the same level
    - `<h3>` (`variant="h3"`): introduces a subsection under the `<h2>` subsection

Headings, and a logical application of heading levels, are indispensable for describing the relationships of belonging that make up the document’s structure. Landmarks and headings together expose the information topology to different users, their assistive software, and agents.

Just because some text appears large or **bold** does not make it a heading. A heading must introduce a section (or subsection) of thematically distinct content. For large (or otherwise attention grabbing) text, not intended as a heading, use a [**Typography**](/docs/components/typography) heading `variant` but change the underlying element using the `render` prop:

```jsx

> <Typography variant="h4" render={<span />}>

Reviewer: This snippet and the preceding paragraph could be moved into a note/caution.

This is also currently the only code snippet under this section that uses `<Typography>`. It's missing an example of using `<Typography>` with heading level customization (e.g. `render={<h2 />}`), which is much more important.

Author: https://github.com/iTwin/stratakit/pull/1443/commits/0b5a78fcc01d17b1fac43373430b8d58ed262a05

---
title: Structure
description: How to successfully organize your content and functionality
---

> Information without structure is just data. High quality interface require logical organization. But structure is perceived and traversed in different ways. The interface one user sees on screen is also parsed by screen reader software, and analyzed by AI agents.

Reviewer 2: interfaces require vs interface requires:

```suggestion
Information without structure is just data. High quality interface requires logical organization. But structure is perceived and traversed in different ways. The interface one user sees on screen is also parsed by screen reader software, and analyzed by AI agents.
```

Author: @GerardasB We're both wrong! It should be "interfaces require" LOL. I'll fix that up in a minute.

Author: https://github.com/iTwin/stratakit/pull/1443/commits/94704b109296598d7f282b5ecfef30b6caff3ed7

In the visual interface, the order of any two objects can appear in three dimensions:

1. Left and right 
2. Above and below
3. Back and front (depth)

Whereas any two components in the source can only be _first and last_:

```jsx
<div>
  <Typography>First object</Typography>
  <Typography>Another object</Typography>
</div>
<Typography>Last object</Typography>
```

Avoid using **CSS** such that it creates a mismatch between visual placement and source order. Use this table to check:

| Where objects are  | Components must be |
| ------------------ | ------------------ |
| left and right     | first and last     |
| above and below    | first and last     |
| back and front     | first and last     |

### Focus order

Interactive elements must be focusable so that the interface can be operated by keyboard. The order in which these elements are encountered and focused must be logical:

| Where buttons are  | Focus must occur  |
| ------------------ | ----------------- |
| left and right     | first then last   |
| above and below    | first then last   |
| back and front     | first then last   |

> Avoid _prioritizing_ the focus of certain elements by applying a positive `tabindex` value or the `autofocus` attribute.

Reviewer 2: Is `tabindex` guidance valid? Should it be geared towards using negative tab index values?

Author: @gerardasB 

* `-1` sometimes (for programmatic/scripted focus)
* `0` occasionally (for making non-native buttons etc focusable—but the native elements are preferred)
* `1+` never: it is a "code smell" that almost invariably leads to focus order (WCAG) issues.

> Avoid _prioritizing_ the focus of certain elements by applying a positive `tabindex` value or the `autofocus` attribute.

Reviewer 2: `autofocus` attribute is specific to dialogs/popovers? `autoFocus` prop (component level) usually does more than just set the attribute?

Author: @GerardasB I wasn't aware of this! The native attribute can be used on any focusable element and is often used on form elements (login screens etc) but `autoFocus` (camel case) behaves differently?

Reviewer 2: Ignore the comment above. 

After glancing through [autofocus](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Global_attributes/autofocus) docs my eyes caught "if it is nested inside a [<dialog>](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/dialog) or [popover](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Global_attributes/popover) element", but missed the "or, ".  What didn't help is that I vaguely remembered that Ariakit did something extra for the `autofocus` attribute: https://ariakit.com/reference/button#autofocus

Reviewer: There are a few different things to understand here:
- The HTML `autofocus` attribute will focus the element on page load, or when a dialog/popover opens.
- The React `autoFocus` prop already diverges from HTML and will focus the element when the component first mounts. However, this is extremely unreliable in practice, especially because components often require multiple renders to fully stabilize.
- Ariakit's `autoFocus` prop attempts to fix the issues with React `autoFocus` by delaying its execution and taking its own dialogs/popups into consideration.
  - Side note: I've wanted us to use Ariakit's [`Focusable`](https://ariakit.com/reference/focusable) (instead of `Role`) for the more common focusable MUI components.

There can be legitimate use cases for setting initial focus. Our guidance should take that into consideration.

```jsx

> <!-- ❌ -->

Reviewer 2: NIT: jsx vs html comments
```suggestion
/* ❌ */
```
or simply `// ❌`

<Link href="/some/page">Focused last</Link>

> <TextField label="Focused first" autofocus />

Reviewer 2: Prop is capitalized
```suggestion
<TextField label="Focused first" autoFocus />
```

<!-- ❌ -->
<ButtonGroup aria-label="Button group">
  <Button>Focused second</Button>

>   <Button tabindex="1">Focused first</Button>

Reviewer 2: Prop is capitalized
```suggestion
  <Button tabIndex="1">Focused first</Button>
```

Author: @gerardasB Damn JSX!

  <Button>Focused last</Button>
</ButtonGroup>
```

More complex interactive components like [**Menu**](/docs/components/menu) and [**Dialog**](/docs/components/dialog) manage focus automatically. Implement these components to avoid focus order issues that could trip up a user or agent.

### Skip link

In the order of major page [landmarks](#landmarks), the `<header>`—typically including a `<nav>`—is expected to come first. 

A skip link is a mechanism for bypassing this header/navigation functionality to interact directly with the main content of the page. It is a provision for keyboard users and is considered an accessibility requirement according to [WCAG](https://www.w3.org/WAI/WCAG22/Understanding/bypass-blocks.html).

Typically, skip links are hidden until focused by keyboard. They are not needed by or available to mouse users.

If you are implementing your own skip link, point the skip link's `href` to the `<main>` element page fragment. To ensure keyboard focus follows the link, include `tabindex="-1"` on the target element.

```html
<a href="#main">skip to content</a>
<!-- page preamble here -->
<main id="main" tabindex="-1">...</main>
```

> For reference, **iTwinUI** has the [SkipToContentLink component](https://itwinui.bentley.com/docs/skiptocontentlink).

Reviewer 2: We should avoid referencing iTwinUI (outside of migration guides)

Author: @GerardasB I've been asked to reference iTwin multiple times in the past, which I always found odd, and I agree it shouldn't be invoked here.

Author: https://github.com/iTwin/stratakit/pull/1443/commits/66900f942582ad09958fd40317e6ddc9f42e3211

Reviewer: Do you have a better reference for a skip link that consumers can look at?

I've tried to negotiate this component being part of our library, but our designers were never interested, so the iTwinUI `SkipToContentLink` was the next best thing.

Reviewer 2: My main issue with linking to iTwinUI is that we are trying to lean towards StrataKit adoption with an end goal where our consumers stop using iTwinUI entirely. Referencing iTwinUI components might be understood as a guidance -> use this specific component which might delay that effort & encourage iTwinUI usage.

An external article maybe? Alternatively, we could implement one as an example?

Author: I agree with Gerardas. It also just muddies the waters further (old Strata, MUI Strata, iTwinUI, "which is it?"). I'll find a decent external article.

EDIT: https://github.com/iTwin/stratakit/pull/1443/commits/0a72af46ae39c78bb15d151c42f32f9f243ebad6

Author: https://github.com/iTwin/stratakit/pull/1443/commits/0a72af46ae39c78bb15d151c42f32f9f243ebad6

  <Button>Underline</Button>
</div>
```

The intent to communicate belonging is clear from the nested structure and the `ButtonGroup` classname. It’s also expressed visually, via CSS, by making the buttons share a common border. However, neither a screen reader nor a crawler will acknowledge the `<div>`—and for good reason! We use `<div>`s to organize our HTML, not the information within it.

To communicate the presence of a group, we can encode it with [`role="group"`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Roles/group_role). Whatever you call this group should be the wording of its accompanying label.

```jsx
<div role="group" aria-label="Text formatting">
  <Button>Bold</Button>
  <Button>Italic</Button>
  <Button>Underline</Button>
</div>
```

In the [**ButtonGroup**](/docs/components/buttongroup) component, **StrataKit** applies `role="group"` automatically. Since the label will be product-specific, you must apply this as the implementer.

The generic `group` role is not the only way to encode a relationship of belonging. If you are not using the following extensively, you are unlikely to be sufficiently communicating structure.

> <table>

Reviewer 2: Is there a reason why the table is authored using HTML rather than md?

Author: @GerardasB Yes, authoring a table of this complexity in markdown syntax is _absolutely horrific_ lmao. But seriously, it would have taken me 3 times as long. Markdown is a superset of HTML, so I see no issue with this syntax here.

Reviewer: This makes the source more difficult to review, and necessitates having to also use `<a>`, `<strong>`, `<code>` and even HTML entities like `&lt;`, instead of simple markdown syntax.

Do you need help converting this table to markdown?

Author: @Mayank99 Be my guest, thank you.

Author: @mayank99 I just tried a couple of html-to-markdown converters to save you the bother. Neither worked, unfortunately. One removed all the table syntax (cheers) and the other removed all _but_ the table syntax (cheers++).

Reviewer: ````md
| Name           | Use case                                                                                                                                                                                                                 | HTML                                                                                                          | Available components                                                                                                                                     |
| -------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Unordered list | Any quantity of items where the order isn’t important, such as checklist items and [Cards](/components/Card)                                                                                                             | The `<ul>` and `<li>` elements or the `role="list"` and `role="listitem"` attribution                         | [List](/components/list)                                                                                                                                 |
| Ordered list   | Any quantity of items where the order is important, such as steps in a process or a breadcrumb trail                                                                                                                     | The `<ol>` and `<li>` elements                                                                                | [Breadcrumbs](/components/breadcrumbs), [Stepper](/components/stepper)                                                                                   |
| Landmark       | Any large region of the page, such as the header, footer, or main content area                                                                                                                                           | `<header>`, `<footer>`, `<main>`, `<nav>`, `<aside>`, `role="search"`                                         | You must implement landmarks as part of your overall page/screen structure. See the [landmarks section](#landmarks).                                     |
| Heading        | Labeling sections and subsections of the page. Heading elements are a special case because they both indicate a section in the document and act as that section’s label. See the [headings](#headings) section for more. | `<h1>` to `<h6>`                                                                                              | The [Typography](/components/typography) component supports the `render` prop, which lets you render the component as a heading element, such as `<h3>`. |
| Fieldset       | Similar to `role="group"` but for grouping form elements, using a `<legend>` for the group’s label                                                                                                                       | `<fieldset>` and `<legend>`                                                                                   |                                                                                                                                                          |
| Figure         | Providing captions for images and other media.                                                                                                                                                                           | `<figure>` and `<figcaption>`. The `<figcaption>` must appear alongside the media, each inside the `<figure>` | None currently provided                                                                                                                                  |
````

Author: https://github.com/iTwin/stratakit/pull/1443/commits/03ffd32d0ec68d4ea5d5962c77828d7de8dd90fa

@mayank99 I dropped this in now. Are we good to go? Just about to post the Language & Labels draft.

Reviewer 2: The table looks good now, resolving. Make sure to install prettier (or run the formatter manually) as there seem to be some formatting inconsistencies.

    <td>You must implement landmarks as part of your overall page/screen structure. See the <a href="#landmarks">landmarks section</a></td>
  </tr>
  <tr>
    <td>Heading</td>
    <td>Labeling sections and subsections of the page. Heading elements are a special case because they both indicate a section in the document and act as that section’s label. See the <a href="#headings">headings</a> section for more.</td>
    <td><code>&lt;h1></code> to <code>&lt;h6></code></td>
    <td>The <a href="/docs/components/typography"><strong>Typography</strong></a> component supports the <code>render</code> prop, which lets you render the component as a heading element, such as <code>&lt;h3></code>.</td>
  </tr>
  <tr>
    <td>Fieldset</td>
    <td>Similar to <code>role="group"</code> but for grouping form elements, using a <code>&lt;legend></code> for the group’s label</td>
    <td><code>&lt;fieldset></code> and <code>&lt;legend></code></td>
    <td></td>
  </tr>
  <tr>
    <td>Figure</td>
    <td>Providing captions for images and other media.</td>

>     <td><code>&lt;figure></code> and <code>&lt;figcaption></code>. The the <code>&lt;figcaption></code> must appear alongside the media, each inside the <code>&lt;figure></code></td>

Reviewer 2: ```suggestion
    <td><code>&lt;figure></code> and <code>&lt;figcaption></code>. The <code>&lt;figcaption></code> must appear alongside the media, each inside the <code>&lt;figure></code></td>
```

    <td>None currently provided</td>
  </tr>
</table>

### Landmarks

[Landmarks](https://www.w3.org/WAI/ARIA/apg/practices/landmark-regions/) are the continents in the map of your interface. Landmarks are large and few in number. Everything (all components) must belong to one landmark or another. They have specific purposes:

> - `<main>`: Include just once per screen/page. This encapsulates the main, unique content of the screen.

Reviewer 2: Would be nice to link to elements/attributes/roles.
```suggestion
- [`<main>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/main): Include just once per screen/page. This encapsulates the main, unique content of the screen.
```

Author: @GerardasB Excellent

Reviewer 2: Un-resolving as code suggestion was only for a single element. There are multiple instances of elements/attributes on the page. At the very least elements in the same list should be linked.

Author: https://github.com/iTwin/stratakit/pull/1443/commits/194b2ca49ebc725bffd70b799b4a1f28896b6449

- `<header>`: Include just one at the start of the document. This may contain a principle `<nav>` element.

> - `<nav>`: Multiple can be used for different purposes, such as site navigation, pagination, and tables of content. Give each `<nav>` a unique label, like `<nav aria-label="table of content">`. This differentiates the landmarks when aggregated in screen reader software.

Reviewer 2: Is TOC grammar correct or should be plural? How about "tables of content"?

```suggestion
- `<nav>`: Multiple can be used for different purposes, such as site navigation, pagination, and tables of content. Give each `<nav>` a unique label, like `<nav aria-label="table of contents">`. This differentiates the landmarks when aggregated in screen reader software.
```

Author: @GerardasB I think either, but this expression gets more search results, so seems to be more prevalent 👍

- `<aside>`: For complementary content and functionality. If your model is the content of `<main>`, then an `<aside>` might contain a tree component for calibrating your model’s layers. Alternatively, the tree may be considered part of the model and the `<aside>` part of the `<main>` content, alongside the model itself.
- `<footer>`: Used once at the end of the document. Useful for including company and product identification, and related links.

### Headings

If landmarks represent continents, then sections are countries: subdivisions of landmarks. Headings _implicitly_ define a section by introducing—rather than wrapping—that section content.

```jsx
<Typography variant="h2">
  Introduction of subsection 1
</Typography>
<p>The content belonging to subsection 1.</p>
```

The provided heading elements, from `<h1>` to `<h6>`, are numbered according to section depth. For example, an `<h3>` following an `<h2>` introduces a subsection to the `<h2>`’s section.

> - `<h1>` (`variant="h1"`): introduces the page

Reviewer 2: I think the basic explanation should be separated more clearly e.g. by simply using h1..h6 elements.
Typography usage/example should be separate and link to Typography component page/docs.

Would tree data-structure be helpful when describing subsections?

Author: https://github.com/iTwin/stratakit/pull/1443/commits/7b47641d410a07cd2065ce7088df0af1bc4de53d

Reviewer 2: `(variant="h1")` should be removed from the `<h1> (variant="h1"): introduces the page`

Additionally, this point of separating general guidance and component usage can be applied to code-snippets under `Belonging` section.

Author: https://github.com/iTwin/stratakit/pull/1443/commits/0093edc400fd9c8fff01fec20cf9704d762934c3
